Std panicking unsafe block in unsafe fn

poliorcetics opened the pull request:
Partial fix of #73904.

This encloses `unsafe` operations in `unsafe fn` in `libstd/ffi/panicking.rs`.

I also made a two lines change to `libstd/thread/local.rs` to add the necessary `unsafe` block without breaking everything else.

@rustbot modify labels: F-unsafe-block-in-unsafe-fn

rust-highfive commented:
r? @shepmaster

(rust_highfive has picked a reviewer for you, use r? to override)

shepmaster commented on src/libstd/thread/local.rs:
```diff
@@ -172,7 +172,8 @@ macro_rules! __thread_local_inner {
                     $crate::thread::__OsLocalKeyInner::new();
 
-                __KEY.get(__init)
+                #[allow(unused_unsafe)]
```
I don't understand why the `allow` is needed.

poliorcetics replied:
The macro is expanded inside the scope of the ˋ#![deny(...)]ˋ, causing an error when the allow is omitted. There are some messages about that in the original issue

shepmaster replied:
What will remind us to remove this `allow` when the call sites are fixed?

poliorcetics replied:
Nothing I know of. As long as the unsafe-op-in-unsafe-fn feature is not both stable and a default, we will need it to allow for both use cases of activating/not activating the feature

Mark-Simulacrum replied:
Hm, I would expect the lint to not propagate into macros, can you file an issue for that? It should probably allow itself inside macros (unless the macro itself denies it).

poliorcetics replied:
See [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8f01524ce1e2668692bc4a7eb4af473c). When activating `#![deny(unsafe_op_in_unsafe_fn)]`, their will be conflicts between places that uses it and places that don't, meaning there is either a missing `unsafe {}` block or an extraneous one. For now, using `#[allow(unused_unsafe)]` is the only solution I know of.

Mark-Simulacrum replied:
Yeah this allow here is fine, for now, but we should tag it with a FIXME pointing at a rust-lang/rust issue, because the allow shouldn't be necessary.

poliorcetics replied:
The issue has been created, see #74838.

Mark-Simulacrum commented on src/libstd/panicking.rs:
```diff
@@ -322,19 +324,36 @@ pub unsafe fn r#try<R, F: FnOnce() -> R>(f: F) -> Result<R, Box<dyn Any + Send>>
+    //
+    // Access to the union's fields: they are made safe by:
+    //  - 1) The call to `into_inner` will allow the data to be dropped later,
+    //    avoiding leaks.
```
This does not belong in the safety comment; leaking memory is not a safety issue. The safety comment should solely be about why this is safe, in this case it's essentially "This is std and std knows that the try intrinsic fills in the `r` or `p` union field based on its return value".

Mark-Simulacrum commented on src/libstd/panicking.rs:
```diff
@@ -322,19 +324,36 @@ pub unsafe fn r#try<R, F: FnOnce() -> R>(f: F) -> Result<R, Box<dyn Any + Send>>
+    //
+    // The call to `intrinsics::r#try` is made safe by:
+    //  - 1) A union is as large as its largest field.
+    //  - 2) See the comment above for the rest of the detailed explanation.
```
These also seem somewhat like they're grasping at something to put rather than being truly helpful, I think we should go hunt down what the try intrinsic's arguments are and figure out appropriate preconditions. In this case I believe those are:
 * do_call, the first argument, can be called with the initial `data_ptr`.
 * do_catch, the second argument, can be called with the `data_ptr` as well.

The do_call and do_catch functions below should also be made unsafe and annotated with the preconditions their bodies imply.

poliorcetics commented:
@Mark-Simulacrum the last commit made the `do_catch` and `do_call` function unsafe, with what I hope are good `SAFETY` comments. I also changed the others to fix them according to your comments.

bors commented:
:umbrella: The latest upstream changes (presumably #73265) made this pull request unmergeable. Please resolve the merge conflicts.

Mark-Simulacrum commented:
r=me with a rebase

poliorcetics commented:
Rebase done, checks are in progress. 😃

Mark-Simulacrum commented:
Please also squash all commits into one.

poliorcetics committed:
Disallow missing unsafe blocks in unsafe fn in panicking.rs

This adds SAFETY comments where necessary, explaining the preconditions
and how they are respected.

Dylan-DPC commented:
@bors r=Mark-Simulacrum rollup

bors commented:
:pushpin: Commit 82ccdab96cd8c146d9958c1050a2d2669dd367e8 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 82ccdab96cd8c146d9958c1050a2d2669dd367e8 -->
<!-- homu: {"type":"Approved","sha":"82ccdab96cd8c146d9958c1050a2d2669dd367e8","approver":"Mark-Simulacrum"} -->
